Wrap errors in order to keep track of their context

chebykin opened the pull request:
When truffle raises an error after interacting with web3.js, it eats the error context. There is no information about a file or line number where did the error occur in my code:

**Current truffle output:**
```sh original output
Contract: error stack traces
       revert opcode
         should fail on revert:
     VM Exception while processing transaction: revert bar -- Reason given: bar.
  RuntimeError: VM Exception while processing transaction: revert bar
      at Function.RuntimeError.fromResults (node_modules/ganache-core/lib/utils/runtimeerror.js:89:13)
      at node_modules/ganache-core/lib/blockchain_double.js:611:28
      at node_modules/ganache-core/node_modules/async/internal/once.js:12:16
      at replenish (node_modules/ganache-core/node_modules/async/internal/eachOfLimit.js:61:25)
      at node_modules/ganache-core/node_modules/async/internal/eachOfLimit.js:71:9
      at eachLimit (node_modules/ganache-core/node_modules/async/eachLimit.js:43:36)
      at node_modules/ganache-core/node_modules/async/internal/doLimit.js:9:16
      at VM.AsyncEventEmitter.emit (node_modules/ganache-core/node_modules/async-eventemitter/lib/AsyncEventEmitter.js:42:3)
      at afterBlock (node_modules/ganache-core/node_modules/ethereumjs-vm/dist/runBlock.js:75:10)
      at node_modules/ganache-core/node_modules/ethereumjs-vm/dist/runBlock.js:252:9
      at node_modules/ganache-core/node_modules/ethereumjs-vm/dist/stateManager.js:387:5
      at node_modules/ganache-core/node_modules/async/dist/async.js:969:16
      at next (node_modules/ganache-core/node_modules/async/dist/async.js:5225:18)
      at node_modules/ganache-core/node_modules/async/dist/async.js:119:16
      at process._tickCallback (internal/process/next_tick.js:61:11)

```

For the moment, the only way to find out the error file/line number is to put a lot of different `console.log('')` and try to figure out the spot between them were did the error occur:

```javascript
it('should do something', async function() {
	console.log('checkpoint-1');
	await myContract.foo();
	console.log('checkpoint-2');
	await myContract.bar();
	console.log('checkpoint-3');
	...
});
```
This is a time westeful and annoying process.

I propose to track an error context somehow. There are lots of places where a context is eaten. The proposed solution handles only several of them. It introduces wrappers which append `original stacktrace` below the initial error message/stacktrace:

```sh
 3) Contract: error stack traces
       revert opcode
         should fail on revert:
     VM Exception while processing transaction: revert bar -- Reason given: bar.
  RuntimeError: VM Exception while processing transaction: revert bar
      at Function.RuntimeError.fromResults (node_modules/ganache-core/lib/utils/runtimeerror.js:89:13)
      at node_modules/ganache-core/lib/blockchain_double.js:611:28
      at node_modules/ganache-core/node_modules/async/internal/once.js:12:16
      at replenish (node_modules/ganache-core/node_modules/async/internal/eachOfLimit.js:61:25)
      at node_modules/ganache-core/node_modules/async/internal/eachOfLimit.js:71:9
      at eachLimit (node_modules/ganache-core/node_modules/async/eachLimit.js:43:36)
      at node_modules/ganache-core/node_modules/async/internal/doLimit.js:9:16
      at VM.AsyncEventEmitter.emit (node_modules/ganache-core/node_modules/async-eventemitter/lib/AsyncEventEmitter.js:42:3)
      at afterBlock (node_modules/ganache-core/node_modules/ethereumjs-vm/dist/runBlock.js:75:10)
      at node_modules/ganache-core/node_modules/ethereumjs-vm/dist/runBlock.js:252:9
      at node_modules/ganache-core/node_modules/ethereumjs-vm/dist/stateManager.js:387:5
      at node_modules/ganache-core/node_modules/async/dist/async.js:969:16
      at next (node_modules/ganache-core/node_modules/async/dist/async.js:5225:18)
      at node_modules/ganache-core/node_modules/async/dist/async.js:119:16
      at process._tickCallback (internal/process/next_tick.js:61:11)
  Original stack: Error
      at TruffleContract.revertable (node_modules/@galtproject/truffle/build/webpack:/packages/truffle-contract/lib/execute.js:160:1)
      at Context.<anonymous> (test/testStorage.js:28:28)
      at process._tickCallback (internal/process/next_tick.js:68:7)
```

In the error stacktrace above `test/testStorage.js:28` points to a particular combination of a file and a line number.

I know this is not a beautiful solution as it modifies default JS error formatting. If it doesn't comply with the project standards, I think we should discuss other ways to solve the problem.

gnidan commented:
This is neat! I'm inclined to merge this, but wondering if we can shorten the stack trace further? Since you manually capture the trace _ahead_ of the exception, the first line (in `lib/execute`) is less information than the second line (`test/testStorage.js`). Really, I'm guessing that the line where it fails in your test is the only information that most people care about, plus the OOG/revert/whatever message.

I would even go so far as to say maybe it'd be worth suppressing the stacktrace from Ganache, and _only_ display the "original". I'll think about this some more, though. What do you think?

Regardless, thanks for opening this!

chebykin commented:
Sure, I think suppressing Ganache stacktraces is a good idea. Also, they take a lot of space too.
In this case, there could be another approach:
* save the current error `stack` fild as another Error object field
* replace the current error `stack` with `originalStackTrace ` (just in case, probably someone needs them for debugging purposes)

Here is an example:
```javascript
catch (e) {
  // not sure about calling it `callStack`
  e.callStack = e.stack;
  e.stack = originalStackTrace;
}
```

gnidan commented:
Oh @chebykin sorry for the delay getting back to you.

I think your plan to overwrite `e.stack` is sound. Go for it, then I think we can look at getting this merged. Thanks!

chebykin commented:
@gnidan, just updated my PR to point to the latest `develop` branch. But there is code duplication now:

* 2 times

```javascript
function appendOriginalStackTrace(e) {
	e.hijackedStack = e.stack;
	e.stack = originalStackTrace.replace(/^Error\n/, e.stack.split('\n')[0]);
	promiEvent.reject(e);
}

```

* and one time

```javascript
web3Error.hijackedStack = web3Error.stack;
web3Error.stack = context.originalStackTrace.replace(/^Error\n/, web3Error.stack.split('\n')[0]);	
```

I'm not sure about what is the best place for keeping these error hijacker methods. Maybe you can suggest me something?

gnidan commented:
Sorry I lost track of this! Adding to my to-dos to circle back here.

gnidan commented:
Yikes, red PR for broken slack notification, silly us!

I see what you mean about the duplication. What do you think of this idea?

1. Define a `lib/promievent.js` inside `@truffle/contract`, containing `class PromiEvent extends Web3PromiEvent`
2. Specify custom logic in the constructor to capture the original stack trace
3. Override the `reject` method to rewrite the stack trace.
4. Find all instances of `Web3PromiEvent` and replace them with the custom class

This way, we only have to define the logic in that one place, but we can be sure it's handled everywhere.

BTW the errors look great!
<img width="866" alt="Screen Shot 2019-09-06 at 8 13 49 PM" src="https://user-images.githubusercontent.com/151065/64467010-e5512f00-d0e2-11e9-9614-ad6c03af2a9a.png">

chebykin commented:
Sure, this is a much cleaner solution.

I have made a wrapper since the original PromiEvent is just a factory, not a constructor. Did I miss something?
